### ISSUE TITLE
Correctly handle Errno::EINTR

### ISSUE THREAD

**Issue by thewoolleyman** [thewoolleyman/process_helper#5]
See http://250bpm.com/blog:12

and https://github.com/ruby/ruby/blob/ruby_2_4/prelude.rb#L36

and https://www.freebsd.org/cgi/man.cgi?read(2)

THis should be captures and just retried in the read nonblock loop.  The only thing that should exit the loop is the EOF exception.

**Comment by thewoolleyman** [comment]
Trying to reproduce on branch: #6

**Comment by thewoolleyman** [comment]
To follow up, I was understanding the problem wrong.

Since my library is a generic library to help execute subprocesses, it should never handle EINTR directly.

Specifically, if EINTR is thrown, as I understand it, it's POSIX telling YOUR process that some signal happened while you were blocked on IO, and you have an opportunity to do any cleanup or finish-up work needed before handling the signal and/or retrying.

Since my library isn't actually doing any direct IO itself, then it should never have any need to handle an EINTR.  If one gets raised, then it was raised by the subprocess.  Maybe correctly, maybe incorrectly, either way, that process is gone and there's nothing I could do about it.

What **IS** the correct thing to do in ruby is to properly handle blocking I/O reads and writes by the subprocess, and retry them once the IO stream is available again.  

This is made easy, because ruby ensures any exception thrown that represents such IO blocking will implement either `IO::WaitReadable` or `IO::WaitWriteable`, and you can be sure that you have handled all possible subprocess IO blocking if you've properly handled those.

So, I [ended up with this](https://github.com/thewoolleyman/process_helper/blob/266fd91557e07dcfc1a92e49f95c2b5193d13391/lib/process_helper.rb#L69):

```
    rescue EOFError
      return output
    rescue IO::WaitReadable
      result = IO.select([stdout_and_stderr], nil, nil, timeout)
      retry unless result.nil?
    rescue IO::WaitWritable
      IO.select(nil, [stdin], nil, timeout)
      retry
    end
```

It took many hours of reading and re-reading man pages, ruby docs, ruby source code to finally realize [what the ruby docs for IO#select](https://ruby-doc.org/core-2.2.2/IO.html#method-c-select) were originally telling me.  But, in my defense I was misled by someone who shall remain nameless who insisted that I should be handling the EINTR ;).  However, I know a lot more about POSIX processes, errors, and signals now :)

**Comment by thewoolleyman** [comment]
Fixed by 266fd91557e07dcfc1a92e49f95c2b5193d13391
